### PR TITLE
DHSCFT-960: Validate uploaded indicator data

### PR DESCRIPTION
**Pull request by ehsh1**
# Description

Jira ticket: https://bjss-enterprise.atlassian.net/browse/DHSCFT-960

Validate uploaded indicator data

## Changes

- Using CSV Helper to validate CSV files

## Validation

Unit tests pass locally

### CODE REVIEW

File: api/DHSC.FingertipsNext.Modules.DataManagement.Service/Validation/UploadedCsvValidator.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        var headerErrors = message.Split('\n').Where(m => m.StartsWith("Header with name", StringComparison.InvariantCulture)).ToList();
+        for (var i = 0; i < headerErrors.Count; i++)
+        {
+            headerErrors[i] = headerErrors.ElementAt(i).Replace("[0]", "", StringComparison.InvariantCulture);
+        }
+
+        return headerErrors;
+    }
+
+    private static CsvConfiguration _setupCsvReaderConfiguration(List<CsvValidationError> errors)
```
[the remark this answers is not in the document]
Comment (ehsh1) [reply api/DHSC.FingertipsNext.Modules.DataManagement.Service/Validation/UploadedCsvValidator.cs:66]: I'll rename to stay consistent with how we already name them

**Review by andyevans-bjss — APPROVED** [review]

File: api/DHSC.FingertipsNext.Modules.DataManagement.Service/Validation/UploadedCsvValidator.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            csv.Context.RegisterClassMap<UploadedIndicatorDataRowMap>();
+            var records = csv.GetRecords<UploadedIndicatorDataRow>();
+
+            try
+            {
+                // Do not call .ToList() on records, as this will bring the entire CSV structure into memory
+                // Turn into a for loop
+                foreach (UploadedIndicatorDataRow row in records)
+                {
+                    Console.WriteLine(row.AreaCode);
```
Comment (gareth-allan-bjss) [inline api/DHSC.FingertipsNext.Modules.DataManagement.Service/Validation/UploadedCsvValidator.cs:35]: I'm not sure we want to be writing the entire contents of the file to the console. Is this because we need to iterate over the rows to trigger validation?

File: api/DHSC.FingertipsNext.Modules.DataManager.UnitTests/Services/Validation/CsvValidationTests.cs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        using (FileStream stream = File.Open(path, FileMode.Open))
+        {
+            result = UploadedCsvValidator.Validate(stream);
+        }
+
+        // Assert
+        result.Success.ShouldBeFalse();
+        result.Errors.Count.ShouldBe(2);
+        result.Errors[0].ErrorMessage.ShouldBeEquivalentTo("The value: '2024-01-AB' is invalid. At row 2 in column 1.");
+        result.Errors[1].ErrorMessage.ShouldBeEquivalentTo("The value: '2024-12-AB' is invalid. At row 3 in column 2.");
```
Comment (gareth-allan-bjss) [inline api/DHSC.FingertipsNext.Modules.DataManager.UnitTests/Services/Validation/CsvValidationTests.cs:74]: Interesting. Do the row numbers start at one and the columns at zero?

File: api/DHSC.FingertipsNext.Modules.DataManagement.Service/Validation/UploadedCsvValidator.cs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+                {
+                    CsvValidationError error = new CsvValidationError(headerError, 1, 1);
+                    errors.Add(error);
+                }
+            }
+        }
+
+        if (errors.Count != 0)
+            return new CsvValidationResult(false, new Collection<CsvValidationError>(errors));
+        else
```
Comment (Mr-Ev-H) [inline api/DHSC.FingertipsNext.Modules.DataManagement.Service/Validation/UploadedCsvValidator.cs:50]: Unnecessary else - just let it drop out to the final return.

File: api/DHSC.FingertipsNext.Modules.DataManagement.Service/Validation/UploadedIndicatorDataRow.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    public double? UpperCi99 { get; set; }
+}
+
+public sealed class UploadedIndicatorDataRowMap : ClassMap<UploadedIndicatorDataRow>
+{
+    const string DateFormatRegex = @"\d{4}-\d{2}-\d{2}";
+
+    public UploadedIndicatorDataRowMap()
+    {
+        Regex dateRegex = new Regex(DateFormatRegex, RegexOptions.CultureInvariant, TimeSpan.FromMilliseconds(100));
```
[the remark this answers is not in the document]
Comment (ehsh1) [reply api/DHSC.FingertipsNext.Modules.DataManagement.Service/Validation/UploadedIndicatorDataRow.cs]: Removed

**Review by gareth-allan-bjss — APPROVED** [review]

**Review by Mr-Ev-H — APPROVED** [review]

File: api/DHSC.FingertipsNext.Modules.DataManager.UnitTests/Services/Validation/CsvValidationTests.cs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+using DHSC.FingertipsNext.Modules.DataManagement.Service.Validation;
+using NSubstitute;
+using Shouldly;
+
+namespace DHSC.FingertipsNext.Modules.DataManagement.UnitTests.Services.Validation;
+
+public class CsvValidationTests
+{
+    /// <summary>
+    /// CSV has invalid columns and data
```
Comment (williamwillis11) [inline api/DHSC.FingertipsNext.Modules.DataManager.UnitTests/Services/Validation/CsvValidationTests.cs:11]: CSV has invalid headers and data

File: api/DHSC.FingertipsNext.Modules.DataManager.UnitTests/Services/Validation/CsvValidationTests.cs [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    /// </summary>
+    [Fact]
+    public void TestReadInvalid()
```
Comment (williamwillis11) [inline api/DHSC.FingertipsNext.Modules.DataManager.UnitTests/Services/Validation/CsvValidationTests.cs:14]: TestReadInvalidHeadersAndInvalidData

File: api/DHSC.FingertipsNext.Modules.DataManager.UnitTests/Services/Validation/CsvValidationTests.cs [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        {
+            result.Errors.ShouldContain(x => x.Value == $"Header with name '{header}' was not found.");
+        }
+    }
+
+    /// <summary>
+    /// CSV has valid columns, but invalid data
+    /// </summary>
+    [Fact]
+    public void TestReadInvalidWithValidHeaders()
```
Comment (williamwillis11) [inline api/DHSC.FingertipsNext.Modules.DataManager.UnitTests/Services/Validation/CsvValidationTests.cs:59]: TestReadValidHeadersAndInvalidData

File: api/DHSC.FingertipsNext.Modules.DataManager.UnitTests/Services/Validation/CsvValidationTests.cs [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        }
+
+        // Assert
+        result.Success.ShouldBeFalse();
+
+        List<string> missingHeaders = new List<string>()
+        {
+            "area_code",
+            "date_from",
+            "date_to",
```
Comment (williamwillis11) [inline api/DHSC.FingertipsNext.Modules.DataManager.UnitTests/Services/Validation/CsvValidationTests.cs:33]: could we have another test case covering where we have a duplicate of one of the headers (but the correct overall number)

so like 'date_from' in their twice instead of 'date_from' and 'date_to'

File: api/DHSC.FingertipsNext.Modules.DataManager.UnitTests/Services/Validation/CsvValidationTests.cs [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        };
+        result.Errors.Count.ShouldBe(missingHeaders.Count);
+        foreach (var header in missingHeaders)
+        {
+            result.Errors.ShouldContain(x => x.Value == $"Header with name '{header}' was not found.");
+        }
+    }
+
+    /// <summary>
+    /// CSV has valid columns, but invalid data
```
Comment (williamwillis11) [inline api/DHSC.FingertipsNext.Modules.DataManager.UnitTests/Services/Validation/CsvValidationTests.cs:56]: CSV has valid headers, but invalid data

File: api/DHSC.FingertipsNext.Modules.DataManager.UnitTests/Services/Validation/CsvValidationTests.cs [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+        CsvValidationResult result;
+        using (FileStream stream = File.Open(path, FileMode.Open))
+        {
+            result = UploadedCsvValidator.Validate(stream);
+        }
+
+        // Assert
+        result.Success.ShouldBeFalse();
+
+        List<string> missingHeaders = new List<string>()
```
Comment (williamwillis11) [inline api/DHSC.FingertipsNext.Modules.DataManager.UnitTests/Services/Validation/CsvValidationTests.cs:29]: could the correct csv headers be defined somewhere and referenced both here and in the implemantation: api/DHSC.FingertipsNext.Modules.DataManagement.Service/Validation/UploadedIndicatorDataRow.cs 
?

File: api/DHSC.FingertipsNext.Modules.DataManager.UnitTests/Services/Validation/CsvValidationTests.cs (hunk 7 above)
Comment (williamwillis11) [reply api/DHSC.FingertipsNext.Modules.DataManager.UnitTests/Services/Validation/CsvValidationTests.cs:14]: and a variation where the csv ONLY has headers with no rows

Comment (williamwillis11) [reply api/DHSC.FingertipsNext.Modules.DataManager.UnitTests/Services/Validation/CsvValidationTests.cs:14]: completely empty csv file ?

File: api/DHSC.FingertipsNext.Modules.DataManager.UnitTests/Services/Validation/CsvValidationTests.cs [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+        result.Errors.Count.ShouldBe(2);
+        result.Errors[0].ErrorMessage.ShouldBeEquivalentTo("The value: '2024-01-AB' is invalid. At row 2 in column 2.");
+        result.Errors[1].ErrorMessage.ShouldBeEquivalentTo("The value: '2024-12-AB' is invalid. At row 3 in column 3.");
+    }
+
+    /// <summary>
+    /// CSV is valid
+    /// </summary>
+    [Fact]
+    public void TestReadValid()
```
[the remark this answers is not in the document]
Comment (gareth-allan-bjss) [reply api/DHSC.FingertipsNext.Modules.DataManager.UnitTests/Services/Validation/CsvValidationTests.cs:82]: What would testing 100 or 1000 rows prove at this unit test level?

**Review by gareth-allan-bjss — APPROVED** [review]

File: api/DHSC.FingertipsNext.Modules.DataManagement.Service/Validation/UploadedCsvValidator.cs [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+{
+    /// <summary>
+    /// Reads the CSV file and attempts to build a list of UploadedIndicatorDataRow
+    /// This will log any validation errors
+    /// </summary>
+    /// <param name="stream"></param>
+    public static CsvValidationResult Validate(Stream? stream)
+    {
+        var errors = new List<CsvError>();
+        if (stream == null)
```
Comment (Mr-Ev-H) [inline api/DHSC.FingertipsNext.Modules.DataManagement.Service/Validation/UploadedCsvValidator.cs:21]: I don't think we should let this method get into the scenario where the stream can be null, I'd suggest stopping the stream from being nullable and handle this further up in the method chain.
